src/v/archival/ntp_archiver_service.cc

```diff
@@ -162,14 +162,19 @@ ss::future<ntp_archiver::batch_result> ntp_archiver::upload_next_candidates(
     for (size_t i = 0; i < _concurrency; i++) {
         vlog(
           ctxlog.debug,
-          "Uploading next candidates for {}, trying offset {}",
+          "Uploading next candidates for {}, trying offset {}, LSO {}",
           _ntp,
-          last_uploaded_offset);
-        auto upload = _policy.get_next_candidate(
-          last_uploaded_offset, high_watermark, lm);
+          last_uploaded_offset,
+          last_stable_offset);
+        auto upload = co_await _policy.get_next_candidate(
+          last_uploaded_offset, last_stable_offset, lm);
         if (upload.source.get() == nullptr) {
             vlog(
-              ctxlog.debug, "Uploading next candidates for {}, ...skip", _ntp);
+              ctxlog.debug,
+              "Uploading next candidates, last_uploaded_offset: {}, "
+              "last_stable_offset: {} ...skip",
+              last_uploaded_offset,
+              last_stable_offset);
```

Comment:
Needs a comment about the nature of the compiler bug

Reply:
done